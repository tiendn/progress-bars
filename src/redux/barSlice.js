import { createSlice } from '@reduxjs/toolkit';
import { notification } from 'antd';

import { getBars } from '../apis/bars';

const defaultState = {
  buttons: [],
  bars: [],
  limit: 9999,
  currentBarIndex: 0,
  loading: true,
};

const barSlice = createSlice({
  name: 'bar',
  initialState: defaultState,
  reducers: {
    showLoading(state) {
      state.loading = true;
    },
    hideLoading(state) {
      state.loading = false;
    },
    fetchBarDataDone(state, action) {
      const { buttons, bars, limit } = action.payload;

      state.buttons = buttons || defaultState.buttons;
      state.bars = bars || defaultState.bars;
      state.limit = limit || defaultState.limit;
    },
    changeActiveBar(state, action) {
      state.currentBarIndex = action.payload.barIndex;
    },
    changeBarValue(state, action) {
      try {
        if (state.bars.length === 0) return;

        const { point } = action.payload;
        const { currentBarIndex, bars } = state;
        let newBarValue = bars[currentBarIndex] + point;

        if (newBarValue < 0) newBarValue = 0;
        state.bars[currentBarIndex] = newBarValue;
      } catch (err) {
        // console.log(err);
        // Report bug
      }
    },
  },
});

export const {
  changeActiveBar,
  changeBarValue,
  fetchBarDataDone,
  showLoading,
  hideLoading,
} = barSlice.actions;

export const getBarData = () => async dispatch => {
  try {
    dispatch(showLoading());
    const data = await getBars();
    dispatch(hideLoading());
    dispatch(fetchBarDataDone({ ...data }));
  } catch (err) {
    dispatch(hideLoading());
    notification.error({ message: err.message });
    console.log(err);
  }
};

export { defaultState };
export default barSlice.reducer;
