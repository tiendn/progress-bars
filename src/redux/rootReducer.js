import { combineReducers } from '@reduxjs/toolkit';
import barReducer from './barSlice';

const rootReducer = combineReducers({
  bar: barReducer,
});

export default rootReducer;
