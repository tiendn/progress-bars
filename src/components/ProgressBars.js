import React from 'react';
import { connect } from 'react-redux';
import { Button, Select } from 'antd';

import BarItem from './BarItem';
import { changeBarValue, changeActiveBar } from '../redux/barSlice';

const { Option } = Select;

const ProgressBars = ({ dispatch, bars, limit, buttons, currentBarIndex }) => {
  const onChangePoint = point => {
    dispatch(changeBarValue({ point }));
  };

  const onChangeCurrentBar = barIndex => {
    dispatch(changeActiveBar({ barIndex }));
  };

  const BarComponents = [];
  const barOptions = [];
  bars.forEach((bar, index) => {
    BarComponents.push(
      <BarItem active={currentBarIndex === index} limit={limit} value={bar} />,
    );
    barOptions.push(<Option value={index}>#progress {index + 1}</Option>);
  });

  const PointButtons = buttons.map(point => (
    <Button
      data-testid="barButton"
      key={point}
      className="barAction"
      onClick={() => onChangePoint(point)}
    >
      {point > 0 ? `+${point}` : point}
    </Button>
  ));

  return (
    <div className="appContainer" data-testid="progressBars">
      <h1>Progress Bars Demo</h1>
      <h3 style={{ color: 'gray' }}>Author: Dao Nam Tien</h3>
      <h3 style={{ color: 'gray', marginBottom: '5%' }}>Limit: {limit}</h3>
      <div className="barContainer">{BarComponents}</div>
      <div className="barActions">
        <Select
          value={currentBarIndex}
          onChange={onChangeCurrentBar}
          placeholder="Please select"
          style={{ marginRight: 10 }}
          className="barAction"
        >
          {barOptions}
        </Select>
        {PointButtons}
      </div>
    </div>
  );
};

const mapStateToProps = ({
  bar: { bars, limit, buttons, currentBarIndex },
}) => ({
  bars,
  buttons,
  currentBarIndex,
  limit,
});

export default connect(mapStateToProps)(ProgressBars);
