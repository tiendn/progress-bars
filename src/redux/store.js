import { configureStore } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import rootReducer from './rootReducer';

const store = configureStore({
  reducer: rootReducer,
});

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./giphySlice', () => {
    const newRootReducer = require('./giphySlice').default;
    store.replaceReducer(newRootReducer);
  });
}

export type AppDispatch = typeof store.dispatch;
export type AppThunk = ThunkAction;
export default store;
