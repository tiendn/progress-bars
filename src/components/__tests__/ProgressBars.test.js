import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import ProgressBars from '../ProgressBars';
import configureStore from '../../redux/store';
import mockData from '../../../mock/data';
import { fetchBarDataDone, changeActiveBar } from '../../redux/barSlice';

describe('ProgressBars testing', () => {
  let store;

  beforeAll(() => {
    store = configureStore;
  });

  it('renders with content', () => {
    const { baseElement } = render(
      <Provider store={store}>
        <ProgressBars />
      </Provider>,
    );

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).not.toBeNull();
  });

  it('change bar value after fire event click button, with current index = 0, value = 59, add +41 with button index = 0', async () => {
    const barIndex = 0;

    store.dispatch(fetchBarDataDone({ ...mockData }));
    store.dispatch(changeActiveBar({ barIndex }));

    const { getAllByTestId } = render(
      <Provider store={store}>
        <ProgressBars />
      </Provider>,
    );

    const BarButtonFirst = await getAllByTestId('barButton')[0];
    const BarItemFirst = await getAllByTestId('barItemValue')[0];

    fireEvent.click(BarButtonFirst);

    // 59 + 41 = 100
    // limit = 170
    expect(BarItemFirst.innerHTML).toBe('58% (100 / 170)');
  });

  it('change bar value after fire event click button, with current index = 0, value = 59, add -47 with button index = 3 two times', async () => {
    const barIndex = 0;

    store.dispatch(fetchBarDataDone({ ...mockData }));
    store.dispatch(changeActiveBar({ barIndex }));

    const { getAllByTestId } = render(
      <Provider store={store}>
        <ProgressBars />
      </Provider>,
    );

    const BarButtonThird = await getAllByTestId('barButton')[3];
    const BarItemFirst = await getAllByTestId('barItemValue')[0];

    fireEvent.click(BarButtonThird);
    fireEvent.click(BarButtonThird);

    // 59 - 47 - 47 < 0
    // limit = 170
    expect(BarItemFirst.innerHTML).toBe('0% (0 / 170)');
  });
});
