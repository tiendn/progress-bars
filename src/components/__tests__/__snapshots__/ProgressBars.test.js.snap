// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBars testing renders with content 1`] = `
<body>
  <div>
    <div
      class="appContainer"
      data-testid="progressBars"
    >
      <h1>
        Progress Bars Demo
      </h1>
      <h3
        style="color: gray;"
      >
        Author: Dao Nam Tien
      </h3>
      <h3
        style="color: gray; margin-bottom: 5%;"
      >
        Limit: 
        9999
      </h3>
      <div
        class="barContainer"
      />
      <div
        class="barActions"
      >
        <div
          class="barAction ant-select ant-select-enabled"
          style="margin-right: 10px;"
        >
          <div
            aria-autocomplete="list"
            aria-controls="test-uuid"
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-select-selection
            ant-select-selection--single"
            role="combobox"
            tabindex="0"
          >
            <div
              class="ant-select-selection__rendered"
            >
              <div
                class="ant-select-selection__placeholder"
                style="display: none; user-select: none;"
                unselectable="on"
              >
                Please select
              </div>
              <div
                class="ant-select-selection-selected-value"
                style="display: block; opacity: 1;"
                title=""
              >
                0
              </div>
            </div>
            <span
              class="ant-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <i
                aria-label="icon: down"
                class="anticon anticon-down ant-select-arrow-icon"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </i>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
